### PR TITLE
Update for Driverkit's new Ubuntu Support

### LINKED ISSUE #28: Update for Driverkit's new Ubuntu Support
**Motivation**

Driverkit PR just merged that updates how it handles Ubuntu (simplifying `ubuntu-aws` and `ubuntu-generic` into just `ubuntu` while also expanding support to the other flavors). Kernel-crawler can now simplify all Ubuntu targets to just `ubuntu`.

**Feature**

* Update all ubuntu targets to just `ubuntu`
* Tangentially, kernel-crawler has been dropping `-generic` from the `kernelrelease` field on `ubuntu-generic` targets. In my experience, `uname -r` output will contain `-generic` on the kernel release. So hopefully this can also be fixed alongside the update. 

**Alternatives**

N/A

**Additional context**

Backwards compatibility should work for now - Driverkit will still support `ubuntu-aws` and `ubuntu-generic` in the meantime, though they should/will be deprecated in future.

### PR DESCRIPTION
**Pull request by EXONER4TED**
Signed-off-by: Logan Bond <lbond@secureworks.com>

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind documentation

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

/area crawler

> /area ci

> /area utils

**What this PR does / why we need it**:

Updates kernel-crawler for Driverkit's new ubuntu support, which simplifies all ubuntu targets to just `ubuntu`. Also update the `ubuntu-generic` targets to include `-generic` on the `kernelrelease`.

**Which issue(s) this PR fixes**:

Fixes #28 

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by FedeDP** [comment]
Hi! Can you share the output from a new Ubuntu run?
At least some lines of course :)

**Comment by EXONER4TED** [comment]
Am literally generating that as we speak 😄 I tested it, then deleted the file, then facepalmed, now regenerating.

**Comment by EXONER4TED** [comment]
It doesn't like that it's json, so had to change to txt:
[ubuntu.txt](https://github.com/falcosecurity/kernel-crawler/files/9151773/ubuntu.txt)

File: kernel_crawler/ubuntu.py [hunk 1]
```diff
@@ -17,19 +17,20 @@ def to_driverkit_config(self, release, deps):
...
+            if 'headers' in dep:
+
+                # set a default flavor
+                flavor = 'generic'
+                # capture the flavor from the string after 'linux-' in the url subdir
                 # example: http://security.ubuntu.com/ubuntu/pool/main/l/linux-oracle/linux-headers-4.15.0-1087-oracle_4.15.0-1087.95_amd64.deb
-                d = re.search(r"(\bl/linux(\-.+\/)?\b)", dep)
-                if d is None:
-                    continue
+                flavor_capture = re.search(r"^.*l/linux-(.+)/.*$", dep)
```
Comment (FedeDP) [inline kernel_crawler/ubuntu.py:26]: Is this an improved version? I mean, shouldn't the old regex work also? It was working before!

**Comment by FedeDP** [comment]
I've got a couple of minor questions: take eg this config:
```
{
      "kernelversion": "113~18.04.1",
      "kernelrelease": "5.4.0-100-hwe-5.4",
      "target": "ubuntu",
      "headers": [
        "http://mirrors.edge.kernel.org/ubuntu/pool/main/l/linux-hwe-5.4/linux-hwe-5.4-headers-5.4.0-100_5.4.0-100.113~18.04.1_all.deb",
        "http://mirrors.edge.kernel.org/ubuntu/pool/main/l/linux-hwe-5.4/linux-headers-5.4.0-100-lowlatency_5.4.0-100.113~18.04.1_amd64.deb",
        "http://mirrors.edge.kernel.org/ubuntu/pool/main/l/linux-hwe-5.4/linux-headers-5.4.0-100-generic_5.4.0-100.113~18.04.1_amd64.deb",
        "http://security.ubuntu.com/ubuntu/pool/main/l/linux-hwe-5.4/linux-headers-5.4.0-100-generic_5.4.0-100.113~18.04.1_amd64.deb",
        "http://security.ubuntu.com/ubuntu/pool/main/l/linux-hwe-5.4/linux-headers-5.4.0-100-lowlatency_5.4.0-100.113~18.04.1_amd64.deb",
        "http://security.ubuntu.com/ubuntu/pool/main/l/linux-hwe-5.4/linux-hwe-5.4-headers-5.4.0-100_5.4.0-100.113~18.04.1_all.deb"
      ]
    },
```

Is kernelrelease `5.4.0-100-hwe-5.4` ok?
And, super minor nitpick (that is not dependent on your PR): can we only store the headers for the specific flavor? Here we've got bot `-hwe` ones, and the `-generic` and the `lowlatency` :/ 

EDIT: regarding this last thing, notice this also:
```
{
      "kernelversion": "113",
      "kernelrelease": "5.4.0-100-generic",
      "target": "ubuntu",
      "headers": [
        "http://mirrors.edge.kernel.org/ubuntu/pool/main/l/linux/linux-headers-5.4.0-100-generic_5.4.0-100.113_amd64.deb",
        "http://mirrors.edge.kernel.org/ubuntu/pool/main/l/linux/linux-headers-5.4.0-100_5.4.0-100.113_all.deb",
        "http://mirrors.edge.kernel.org/ubuntu/pool/main/l/linux/linux-headers-5.4.0-100-lowlatency_5.4.0-100.113_amd64.deb",
        "http://security.ubuntu.com/ubuntu/pool/main/l/linux/linux-headers-5.4.0-100_5.4.0-100.113_all.deb",
        "http://security.ubuntu.com/ubuntu/pool/main/l/linux/linux-headers-5.4.0-100-generic_5.4.0-100.113_amd64.deb",
        "http://security.ubuntu.com/ubuntu/pool/main/l/linux/linux-headers-5.4.0-100-lowlatency_5.4.0-100.113_amd64.deb"
      ]
    },
```

This is the same config above but for -generic instead of -hwe.  But notice that the "_all" headers here has no "-$flavor" suffix.

**Comment by EXONER4TED** [comment]
`5.4.0-100-hwe-5.4` is okay - the flavor will end up being `hwe-5.4` in Driverkit, which... if you look at the Driverkit tests, is an expected/accepted input: https://github.com/falcosecurity/driverkit/blob/master/pkg/driverbuilder/builder/ubuntu_test.go#L104-L113

(actually I made those tests from kernel-crawler generated output from the `main` branch) - funny enough, I couldn't get Driverkit to find headers for releases like that and had to rely on kernel crawler to provide them. 

That specific config was actually build successfully in my local testing of the ubuntu support update over in Driverkit:
```shell
╰─❯ ll output/scwx_ubuntu-hwe-5.4_5.4.0-100-hwe-5.4_113~18.04.1.ko
-rw-r--r-- 1 lbond 682K Jul 15 22:32 output/scwx_ubuntu-hwe-5.4_5.4.0-100-hwe-5.4_113~18.04.1.ko
```

To answer your second question, yea... I don't know why kernel-crawler returns lowlatency, hwe, generic, etc. all for the same hwe flavor. I think this goes back to how hwe kernels appear to be stored differently. Another engineer in our org noticed that hwe kernels tend to map to LTS kernels, and maybe that is why? Maybe some further investigation needs to be done there.

In regards to your last point where the `_all` package is missing a flavor in the URL... that's interesting. Let me see what Driverkit does with that if I have it build.

File: kernel_crawler/ubuntu.py (hunk 1 above)
Comment (EXONER4TED) [reply kernel_crawler/ubuntu.py:26]: The previous regex included the `-` in the capture group and the ending `/`. To get around that problem, the code also previously was doing this to get the kernel release:
```python
                d = d.group(0)
                 target = "ubuntu" + d[7:len(d) - 1]
                 release = krel + d[7:len(d) - 1]
```
...which is a bit hard to follow in code (where did that 7 even come from?)

So I simplified it to exclude the `-` and the `/` from the capture group and actually just capture the flavor directly. Then we do not need to use array slicing to reliably get the flavor.

The previous regex also used: `(\b` and `\b)` to start and end the expression, which I just didn't recognize. Using `^.*` and `.*$`is more generic and explicitly defines the start/end of the expression.

Comment (FedeDP) [reply kernel_crawler/ubuntu.py:26]: Uh thank you! Yep it looks really better eheh
thanks for the kind explanation!

**Comment by EXONER4TED** [comment]
Just ran Driverkit with config:
```yaml
kernelversion: "113"
kernelrelease: "5.4.0-100-generic"
target: "ubuntu"
kernelurls:
  - "http://mirrors.edge.kernel.org/ubuntu/pool/main/l/linux/linux-headers-5.4.0-100-generic_5.4.0-100.113_amd64.deb"
  - "http://mirrors.edge.kernel.org/ubuntu/pool/main/l/linux/linux-headers-5.4.0-100_5.4.0-100.113_all.deb"
  - "http://mirrors.edge.kernel.org/ubuntu/pool/main/l/linux/linux-headers-5.4.0-100-lowlatency_5.4.0-100.113_amd64.deb"
  - "http://security.ubuntu.com/ubuntu/pool/main/l/linux/linux-headers-5.4.0-100_5.4.0-100.113_all.deb"
  - "http://security.ubuntu.com/ubuntu/pool/main/l/linux/linux-headers-5.4.0-100-generic_5.4.0-100.113_amd64.deb"
  - "http://security.ubuntu.com/ubuntu/pool/main/l/linux/linux-headers-5.4.0-100-lowlatency_5.4.0-100.113_amd64.deb"
output:
  module: "/tmp/test.ko"
```
and it worked just fine. I suppose it's because kernel crawler is explicitly telling Driverkit where to look, and Driverkit actually just `curl`'s and unpacks whatever URLs you give it.

BUT, if I remove the `kernelurls` option from Driverkit, it is unable to find the headers. So this is another edge case we may want to update Driverkit to handle if crawling in Driverkit is still considered core functionality (I would maybe argue that we could just rely on kernel-crawler for that now).

From the debug output though, Driverkit is setting the `sourcedir` for the kernel headers to this:
```
sourcedir=/tmp/kernel-download/usr/src/linux-headers-5.4.0-100-generic 
```
Which I believe we would call correct, yes? That's because the regex in Driverkit is doing `linux-headers*{flavor}`, which is correct in this case for `linux-headers*generic`.

What do you think?

**Comment by EXONER4TED** [comment]
I guess a good question to answer as well is do we want to return `lowlatency` and `hwe` urls with generic kernels...? Or is that a bug in kernel-crawler?

I think Driverkit is agnostic since it `find`s the correct directory with `KernelHeadersPattern`.

**Comment by FedeDP** [comment]
> BUT, if I remove the kernelurls option from Driverkit, it is unable to find the headers. So this is another edge case we may want to update Driverkit to handle if crawling in Driverkit is still considered core functionality (I would maybe argue that we could just rely on kernel-crawler for that now).

Yeah, the internal driverkit logic to crawl kernels will be kept as a fallback, but the preferred way is using configs with `kernelurls` provided!

> Which I believe we would call correct, yes? That's because the regex in Driverkit is doing linux-headers*{flavor}, which is correct in this case for linux-headers*generic.

You're right, i forgot about the regex!

> I guess a good question to answer as well is do we want to return lowlatency and hwe urls with generic kernels...? Or is that a bug in kernel-crawler?

I think kernel-crawler should not return lowlatency and hwe urls; i am not sure whether it is a bug in the kernel-crawler or not! :/

**Comment by EXONER4TED** [comment]
Let me see what I can do there. It's a bit out of scope for this PR, but maybe it's a simple fix to just filter the URLs returned to `if flavor in url` or something...

**Comment by FedeDP** [comment]
Well given that driverkit works fine in any case, i think we can avoid wasting too much time on that now :) 
Thank you btw!

**Review by FedeDP — APPROVED** [review]
/approve

**Comment by FedeDP** [comment]
Hold until a new Driverkit version is tagged and updated on test-infra!

File: kernel_crawler/ubuntu.py (hunk 1 above)
Comment (maxgio92) [reply kernel_crawler/ubuntu.py:26]: Good catch @EXONER4TED! thank you. We could also leave the word boundaries, with something like `\bl/linux-(.+)/\b` as we don't need to catch start and end matches.

LGTM, I don't get how it worked before :/

**Comment by FedeDP** [comment]
/unhold

We can start merging this; new driverkit is out with its support.
We will wait on test-infra to be ready before merging any kernels lists update!
